pxr/usd/usdSemantics/labelsQuery.cpp

```diff
@@ -0,0 +1,229 @@
+//
+// Copyright 2024 Pixar
+//
+// Licensed under the Apache License, Version 2.0 (the "Apache License")
+// with the following modification; you may not use this file except in
+// compliance with the Apache License and the following modification to it:
+// Section 6. Trademarks. is deleted and replaced with:
+//
+// 6. Trademarks. This License does not grant permission to use the trade
+//    names, trademarks, service marks, or product names of the Licensor
+//    and its affiliates, except as required to comply with Section 4(c) of
+//    the License and to reproduce the content of the NOTICE file.
+//
+// You may obtain a copy of the Apache License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the Apache License with the above modification is
+// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. See the Apache License for the specific
+// language governing permissions and limitations under the Apache License.
+//
+
+#include "pxr/usd/usdSemantics/labelsQuery.h"
+
+#include <mutex>
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+UsdSemanticsLabelsQuery::UsdSemanticsLabelsQuery(const TfToken& taxonomy,
+                                                 UsdTimeCode timeCode) :
+    _taxonomy(taxonomy), _time(timeCode) {
+    TF_VERIFY(!taxonomy.IsEmpty());
+}
+
+UsdSemanticsLabelsQuery::UsdSemanticsLabelsQuery(const TfToken& taxonomy,
+                                                 const GfInterval& interval) :
+    _taxonomy(taxonomy), _time(interval) {
+    TF_VERIFY(!taxonomy.IsEmpty());
+    if (!TF_VERIFY(!interval.IsEmpty())) {
+        _time = UsdTimeCode::Default();
+    }
+}
+
+bool
+UsdSemanticsLabelsQuery::_PopulateLabels(const UsdPrim& prim) {
+    TF_DEV_AXIOM(!prim.IsPseudoRoot());
+    UsdSemanticsLabelsAPI schema(prim, _taxonomy);
+    if (!schema.GetPrim().HasAPI<decltype(schema)>(schema.GetName()))
```

Comment:
Constructing a schema object is not trivial in cost; can we do this check first, using the given `prim`, `UsdSemanticsLabelsAPI`, and `_taxonomy` ?

Reply:
I don't think we should be doing this check at all-- we should just be validating the schema.

Reply:
The HasAPI check is a faster check, though, and given that we expect, in many scenes, to have few if any applications of the schema, it may be worth doing the fast check up front?